File: 3d/src/main/java/io/github/xerooup/frame3d/audio/AudioManager.java
package io.github.xerooup.frame3d.audio;

import org.lwjgl.openal.AL;
import org.lwjgl.openal.ALC;
import org.lwjgl.openal.ALC10;
import org.lwjgl.openal.AL10;

import java.nio.ByteBuffer;
import java.nio.IntBuffer;
import java.util.ArrayList;
import java.util.List;

public class AudioManager {
    private static long device = 0;
    private static long context = 0;
    private static boolean initialized = false;
    private static final List<Sound> sounds = new ArrayList<>();

    private static float masterVolume = 1.0f;

    private AudioManager() {} // prevent instantiation

    // initializes openal audio system
    public static void init() {
        if (initialized) return;

        // open default audio device
        device = ALC10.alcOpenDevice((ByteBuffer) null);
        if (device == 0L) {
            throw new IllegalStateException("failed to open OpenAL device");
        }

        // create audio context
        context = ALC10.alcCreateContext(device, (IntBuffer) null);
        if (context == 0L) {
            ALC10.alcCloseDevice(device);
            throw new IllegalStateException("failed to create OpenAL context");
        }

        if (!ALC10.alcMakeContextCurrent(context)) {
            ALC10.alcDestroyContext(context);
            ALC10.alcCloseDevice(device);
            throw new IllegalStateException("failed to make OpenAL context current");
        }

        // create openal capabilities
        AL.createCapabilities(ALC.createCapabilities(device));
        initialized = true;
    }

    // gets master volume (0.0 to 1.0)
    public static float getMasterVolume() {
        return masterVolume;
    }

    // sets master volume (0.0 to 1.0)
    public static void setMasterVolume(float volume) {
        masterVolume = Math.max(0f, Math.min(1f, volume));
        // update volume for all registered sounds
        for (Sound sound : sounds) {
            sound.updateVolume();
        }
    }

    // registers a sound for global management
    static void registerSound(Sound sound) {
        sounds.add(sound);
    }

    // unregisters a sound
    static void unregisterSound(Sound sound) {
        sounds.remove(sound);
    }

    // stops all playing sounds
    public static void stopAll() {
        for (Sound sound : sounds) {
            sound.stop();
        }
    }

    // pauses all playing sounds
    public static void pauseAll() {
        for (Sound sound : sounds) {
            if (sound.isPlaying()) {
                AL10.alSourcePause(sound.getSourceId());
            }
        }
    }

    // resumes all paused sounds
    public static void resumeAll() {
        for (Sound sound : sounds) {
            int state = AL10.alGetSourcei(sound.getSourceId(), AL10.AL_SOURCE_STATE);
            if (state == AL10.AL_PAUSED) {
                AL10.alSourcePlay(sound.getSourceId());
            }
        }
    }

    // cleans up all audio resources
    public static void dispose() {
        if (!initialized) return;

        stopAll();
        sounds.clear();

        if (context != 0L) {
            ALC10.alcMakeContextCurrent(0L);
            ALC10.alcDestroyContext(context);
            context = 0L;
        }

        if (device != 0L) {
            ALC10.alcCloseDevice(device);
            device = 0L;
        }

        initialized = false;
    }

    // returns current master volume multiplier
    static float getVolumeMultiplier() {
        return masterVolume;
    }

    // checks if audio system is initialized
    public static boolean isInitialized() {
        return initialized;
    }
}

File: 2d/src/main/kotlin/io/github/xerooup/frame2d/audio/AudioManager.kt
package io.github.xerooup.frame2d.audio

import org.lwjgl.openal.AL
import org.lwjgl.openal.ALC
import org.lwjgl.openal.ALC10
import org.lwjgl.openal.AL10
import java.nio.ByteBuffer
import java.nio.IntBuffer

object AudioManager {
    private var device: Long = 0
    private var context: Long = 0
    var initialized = false
    private val sounds = mutableListOf<Sound>()

    // global volume control (0.0 to 1.0)
    var masterVolume: Float = 1.0f
        set(value) {
            field = value.coerceIn(0f, 1f)
            // update volume for all registered sounds
            sounds.forEach { it.updateVolume() }
        }

    // initializes openal audio system
    fun init() {
        if (initialized) return

        // open default audio device
        device = ALC10.alcOpenDevice(null as ByteBuffer?)
            ?: throw IllegalStateException("failed to open OpenAL device")

        // create audio context
        context = ALC10.alcCreateContext(device, null as IntBuffer?)
        ALC10.alcMakeContextCurrent(context)

        // create openal capabilities
        AL.createCapabilities(ALC.createCapabilities(device))
        initialized = true
    }

    // registers a sound for global management
    internal fun registerSound(sound: Sound) {
        sounds.add(sound)
    }

    // unregisters a sound
    internal fun unregisterSound(sound: Sound) {
        sounds.remove(sound)
    }

    // stops all playing sounds
    fun stopAll() {
        sounds.forEach { it.stop() }
    }

    // pauses all playing sounds
    fun pauseAll() {
        sounds.forEach { if (it.isPlaying()) AL10.alSourcePause(it.getSourceId()) }
    }

    // resumes all paused sounds
    fun resumeAll() {
        sounds.forEach { if (AL10.alGetSourcei(it.getSourceId(), AL10.AL_SOURCE_STATE) == AL10.AL_PAUSED) {
            AL10.alSourcePlay(it.getSourceId())
        } }
    }

    // cleans up all audio resources
    fun dispose() {
        if (!initialized) return
        stopAll()
        sounds.clear()
        ALC10.alcDestroyContext(context)
        ALC10.alcCloseDevice(device)
        initialized = false
    }

    // returns current master volume multiplier
    internal fun getVolumeMultiplier(): Float = masterVolume
}